Buildcache pruning

mvandenburgh opened the pull request:
This PR implements buildcache pruning. The pruning logic is a modified version of the "Orphaned Pruning" algorithm described [here](https://github.com/spack/spack-infrastructure/tree/main/images/ci-prune-buildcache#orphan-pruning). It works as follows:
1. User runs `spack buildcache prune <mirror>`
2. Fetch all manifests and their associated metadata from the cache
3. Fetch a list of blobs present in the cache
4. Diff these two sets to identify manifests that reference non-existant blobs (**orphaned manifests**) and blobs that are not referenced by any manifests (**orphaned blobs**) and delete them
5. Repeat steps 2-4 until there are no orphaned manifests or blobs found

The `buildcache prune` command is intentionally barebones and not very configurable; the intent is to get a simple pruning approach implemented into spack for now, and follow up with more configuration options (allowing users to specify individual specs to prune, other pruning algorithms, etc.) later.

mvandenburgh committed:
Implement orphan pruning

mvandenburgh committed:
Tests for pruning

mvandenburgh committed:
Parallelize pruning object fetching/deleting

mvandenburgh committed:
Update shell completions

mvandenburgh commented:
@spackbot fix style

mvandenburgh committed:
[@spackbot] updating style on behalf of mvandenburgh

mvandenburgh committed:
Fix type checking errors

mvandenburgh committed:
Fix flake8 errors

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,155 @@
+                if cache_entry:
+                    cache_entry.destroy()
+
+        with ThreadPoolExecutor() as executor:
```
Can you use `spack.util.parallel.make_concurrent_executor()` here instead?  For one thing it creates a `ProcessPoolExecutor`, which is important since I think it's still true that the urllib spack uses is not thread-safe.  That function also falls back to serial execution on platforms that don't have forking.

mvandenburgh replied:
Fixed. I still kept the usage of `concurrent.futures.as_completed()`, which should work since `spack.util.parallel.make_concurrent_executor()` returns an instance of `concurrent.futures.Executor`. Is that ok? I couldn't find examples of this pattern elsewhere in the codebase.

We could also just iterate over the list of `Future` objects and call `.result()` on each, but that would be less efficient since it would wait for each `Future` to resolve sequentially.

scottwittenburg replied:
No, I think `as_completed()` is fine here, better even.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,155 @@
+            tty.warn(f"Unable to prune blob {url} due to: {e}")
+            return 0
+
+    with ThreadPoolExecutor() as executor:
```
Same here, `spack.util.parallel.make_concurrent_executor()`

scottwittenburg commented on lib/spack/spack/cmd/buildcache.py:
```diff
@@ -211,6 +212,10 @@ def setup_parser(subparser: argparse.ArgumentParser):
     download.set_defaults(func=download_fn)
 
+    prune = subparsers.add_parser("prune", help=prune_fn.__doc__)
+    prune.add_argument("mirror", type=arguments.mirror_name, help="name of a configured mirror")
```
Can you follow the example [here](https://github.com/spack/spack/blob/dade498f7237fd5104fbf82a5f91dbc95ab089d9/lib/spack/spack/cmd/buildcache.py#L75)? It will allow users more flexibility in how they specify the target mirror.

scottwittenburg reviewed (changes requested):
This is coming along nicely @mvandenburgh, thanks! I've requested a couple concrete changes, but I'm also wondering if it would be possible to avoid re-fetching all the manifests and re-listing all the blobs each time through the top-level loop. I think I see how one pass of the algorithm could create new orphans, thus creating the need to repeat until there are no orphans. But it seems like it should be possible to achieve by just fetching the manifests and listing the blobs once, assuming some extra bookkeeping.  Feel free to ping me if you want to meet and chat about it.

mvandenburgh committed:
Refactor to avoid fetching manifests on every loop

This commit moves the fetching of manifests into a context manager
that wraps the entire loop. For a buildcache with *n* manifests, this
change reduces the number of manifest fetches from *n* to 1.

mvandenburgh committed:
Use spack.util.parallel.make_concurrent_executor

mvandenburgh committed:
Fix `mirror` CLI arguemnt

mvandenburgh committed:
Add dry-run mode to prune cmd

mvandenburgh committed:
Remove parallelization of manifest fetching

The current manifest-fetching API requires calling a function
for each manifest that needs to be processed. Because Python functions
cannot be serialized to other processes, we cannot parallelize
this.

mvandenburgh committed:
Move inlined functions into global scope

Inline functions can't be used in a `ProcessPoolExecutor` context.

mvandenburgh committed:
Fix infinite loop in dry-run mode

mvandenburgh committed:
Update tests

mvandenburgh committed:
Update shell completions

mvandenburgh committed:
Regenerate shell completions outside of spack environment

scottwittenburg commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -748,3 +750,108 @@ def test_migrate_requires_index(capsys, v2_buildcache_layout, mutable_config):
+    manifest_url = f"file://{cache_entry.get_manifest_url(spec=spec, mirror_url=mirror_directory)}"
+
+    # Remove the blobs from the cache, orphaning the manifest
+    for blob_file in cache_entry.read_manifest().data:
```
Maybe no need to `read_manifest()` again here?

scottwittenburg commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -748,3 +750,108 @@ def test_migrate_requires_index(capsys, v2_buildcache_layout, mutable_config):
+    cache_entry = URLBuildcacheEntry(
+        mirror_url=f"file://{mirror_directory}", spec=spec, allow_unsigned=True
+    )
+    cache_entry.read_manifest()
```
`read_manifest()` here returns the manifest, so you could just store it in a variable and not have to read it again below.

mvandenburgh committed:
Remove redundant usage of `capsys`

scottwittenburg commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -748,3 +750,101 @@ def test_migrate_requires_index(capsys, v2_buildcache_layout, mutable_config):
+        # should raise an error because it has been pruned
+        with pytest.raises(BuildcacheEntryError):
+            cache_entry.manifest = None
+            cache_entry.read_manifest()
```
I don't think I was considering this case when writing the `URLBuildCacheEntry` class, though I guess it's reasonable behavior and works as you've discovered.

Fwiw, I would have thought that just the existence check above is sufficient, but let me know if you had something more in mind here.

mvandenburgh replied:
Now that I'm looking at this again, it seems like this case is actually testing the behavior of the `URLBuildCacheEntry` class (and not pruning) since, as you said, we've already checked for the existence of it at this point. 
I'm now leaning toward removing this. Another benefit of removing it is that would mean that `if dry_run` conditional right above it goes away (I thought that was a little ugly since `dry_run` is a pytest parameter)

mvandenburgh replied:
https://github.com/spack/spack/pull/50778/commits/1af4159a48161d67cf33c3fc5ffe9b282b29f3de

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,197 @@
+                }
+            )
+        except Exception as e:
+            tty.warn(f"Unable to fetch spec for manifest {manifest} due to: {e}")
```
Is this the right error message?  Seems to me what could have failed is reading the manifest itself, nothing to do with specs.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,197 @@
+                cache_entry.destroy()
+
+    # Blobs that are referenced in a manifest file (but not necessarily present in the cache)
+    blob_hashes_referenced_by_manifest = set(blob_to_manifest_mapping.keys())
```
nit: This isn't actually a set of hashes, is it? It looks to me like it actually contains blob urls.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,197 @@
+
+    :param mirror: The mirror from which to fetch the manifests.
+    :return: A list of tuples, each containing a list of file names and a callable to read
+             the manifest entries.
```
I see this is an internal method, but as long as it has a docstring, it could be updated to reflect that it lists the blobs and returns that list in the tuple.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,197 @@
+    blob_hashes_referenced_by_manifest = set(blob_to_manifest_mapping.keys())
+
+    # Blobs that are actually present in the cache (but not necessarily referenced in any manifest)
+    blob_hashes_present_in_cache: Set[str] = set(blobs)
```
nit: Again not hashes, but urls, or am I misreading?

mvandenburgh committed:
Remove unnecessary delete

The `blob_to_manifest_mapping` dict was previously being used for some
additional bookkeeping, but not anymore. We don't need to worry about
removing keys from it as blobs are deleted.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,197 @@
+        for manifest in orphaned_manifests:
+            futures.append(executor.submit(_delete_object, manifest, dry_run))
+            try:
+                manifests.remove(manifest)
```
My feeling is this is a bug in the `aws` case, since in that case, these manifest urls are actually pointing to the temp directory on the filesystem. Deleting those wouldn't actually delete the manifest from the mirror, would it?  Maybe I'm missing something.

mvandenburgh replied:
I've added logic to use the `aws s3 sync` command to handle manifest deletion if it's available.

scottwittenburg reviewed (changes requested):
Mostly tiny things this time around, except maybe one big thing.  Let me know if I misread the code or something though.  Thanks @mvandenburgh!

mvandenburgh committed:
Properly delete manifests when fetched w/ aws cli

mvandenburgh committed:
Remove redundant `read_manifest`

mvandenburgh committed:
Improve docstring/variable names

Updates the return type in the docstring to be correct, and also
renames a variable to be more descriptive.

mvandenburgh committed:
Fix inaccurate variable names

These sets contain blob *urls*, not hashes.

mvandenburgh committed:
Fix incorrect error messages

mvandenburgh committed:
Remove redundant test assertion

scottwittenburg commented on lib/spack/spack/test/cmd/buildcache.py:
```diff
@@ -748,3 +749,94 @@ def test_migrate_requires_index(capsys, v2_buildcache_layout, mutable_config):
+    cache_entry = URLBuildcacheEntry(
+        mirror_url=f"file://{mirror_directory}", spec=spec, allow_unsigned=True
+    )
+    cache_entry.fetch_metadata()
```
Just to make sure I'm not missing something, is there a reason to `fetch_metadata()` here? It doesn't seem like you use the fetched file anywhere.  If you don't need it, you could just say `manifest = cache_entry.read_manifest()`, like you do in the test below. And then you don't need the `cache_entry.destroy()` at the bottom.

mvandenburgh replied:
I did that so that `cache_entry.manifest` is populated by the time I refer to it in the next line to construct the `blob_urls` list. But I didn't realize at the time you can just call `.read_manifest()` directly, so you're correct that it's not necessary. See https://github.com/spack/spack/pull/50778/commits/caeaf58b168852e75db16f7e8c54dad5823e60fe

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,274 @@
+    if dry_run:
+        pruned_object_count = len(orphaned_blobs) + len(orphaned_manifests)
+        for manifest in orphaned_manifests:
+            manifests.remove(manifest)
```
This could be something for future work: Here it would be nice to indicate the things that would be removed in the `--dry-run`, as without it, all you get is a summary of the number of things that would be deleted.  I was at first confused when I did a dry run and saw more items than I expected (which turned out to be correct, just not initially obvious to me).

mvandenburgh replied:
It wasn't that much work, and I agree it makes this command much more useful, so I went ahead and added it.

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,274 @@
+    total_pruned = 0
+    with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpspecsdir:
+        manifest_list, read_fn, blob_list = _fetch_manifests(mirror, tmpspecsdir)
+        while True:
```
This could be for the future: If we find this iterative approach isn't performant enough, maybe we can circle back and try to handle it in a single pass by doing some extra book-keeping.

mvandenburgh replied:
Agreed. Since it would complicate things/require reworking how we differentiate between deletion via `aws s3 sync` and the "manual" way w/ `web_util`, I vote for future :slightly_smiling_face:

scottwittenburg commented on lib/spack/spack/buildcache_prune.py:
```diff
@@ -0,0 +1,274 @@
+
+        tty.info(
+            ("Would have pruned" if dry_run else "Pruned")
+            + f" {total_pruned} orphaned objects from mirror: {mirror.fetch_url}"
```
I'm kind of thinking it would be nice, if this wasn't a `--dry-run` and anything was pruned, to at least tell the user that the buildcache index is likely out of date now and should be regenerated.

scottwittenburg reviewed (changes requested):
Thanks @mvandenburgh this is getting close.  I've added a couple comments that could just be addressed sometime in the future, one that I think would be good to do on this PR (warning to update buildcache index), and one thing you can do now, later, or ignore (in the test method).

mvandenburgh committed:
Remove unnecessary `fetch_metadata()` call

mvandenburgh committed:
Notify user index is likely outdated after pruning

mvandenburgh committed:
Log out every prunable manifest/blob if running in dry-run

mvandenburgh committed:
Style

scottwittenburg reviewed (approved):
I took it for a test drive and it seems to behave like we have discussed, thanks @mvandenburgh!
